pkg/clusteroperator/clusteroperator.go

```diff
@@ -104,12 +105,23 @@ func (s *StatusHandler) Update(condtype osapi.ClusterStatusConditionType, status
 			LastTransitionTime: metaapi.Now(),
 		})
 
-		desiredVersions := []osapi.OperandVersion{
-			{
-				Name:    "operator",
-				Version: version.Version,
-			},
+		// when we are at the available level, reset versions in status to the input
+		// release version
+		desiredVersions := state.Status.Versions
+		if releaseVersion := os.Getenv("RELEASE_VERSION"); len(releaseVersion) > 0 {
+			// an available operator resets release version
+			if condtype == osapi.OperatorAvailable && status == osapi.ConditionTrue {
+				desiredVersions = []osapi.OperandVersion{
+					{
+						Name:    "operator",
+						Version: releaseVersion,
+					},
+				}
+			}
+		} else {
+			desiredVersions = nil
```

Comment:
So to make sure i understand..this logic means that basically as long as the registry is considered available (pods running, regardless of image version), the operator is going to report its version as being the current operator version.

Meaning when we go from registry operator v4.0 to registry operator v4.1, the operator will basically immediately report itself as being at v4.1 (assuming the registry was available already).

At some point subsequent to that, the registry operator will redeploy the registry (to move it to the new v4.1 image), at which point the registry might go unavailable(if for some reason it can't come up), at which point the registry operator will still report being at v4.1, but not unavailable.

Is that our meets min here, or do we need tighter logic in the registry operator that ensures the registry pods are running at the new version before the operator reports itself as being at the new version?

Reply:
That doesn't meet the min.  I'm trying to interpret how your operator knows that it has reached level.

Reply:
right now i'd say it probably does not know, at least not in a tightly coupled way.

It knows it has updated the deployment to use the new image, and it knows whether the deployment is available or not.  But it doesn't know whether all pods in the deployment are running the new image level or not.  That will require more logic around looking at the deployment.

at a minimum your logic above should be gating on available==true && progressing==false.  That still doesn't get us 100% of the way there i don't think, but it's closer to reflecting the reality of "we have achieved the new level"

Reply:
Actually, looking at operator/controller it does the following:

1. update the deployment
2. load the deployment object
3. pass the deployment object to syncStatus
4. check if the deployment is available
5. sets available true if the deployment is available

The condition that needs to be checked is `isDeploymentStatusAvailable` but with ensuring the generation is updated (like the CVO does).

I believe with that change it would be sufficient to state that you are now available for the version.

Reply:
> The condition that needs to be checked is `isDeploymentStatusAvailable` but with ensuring the generation is updated (like the CVO does)

yeah that's roughly what was in my head.  However we are still "available=true" even if the generation is not updated.  however we would be progressing=true.

we need to:

1) mark ourselves progressing=true and some marker about the generation the deployment is at so we know when the deployment achieves the new generation
2) update the deployment to the new image
3) wait for the deployment achieved generation to be updated
4) now we can mark progressing=false and set the new version

Reply:
we don't really need the marker - it all happens in the same sync loop.  we only need to reset the versions when we hit "level" and what we should really be doing is avoiding changing the available condition status (skip updateCondition) when we have available replicas but it's not up to date

Reply:
Updated with a slight tweak